### ISSUE TITLE
DB open & table being created but erroring out as 'table doesn't exist'

### ISSUE THREAD

**Issue by rolinger** [brody2consult/cordova-sqlite-storage#451]
/\* three issues:
1.  everytime I execute the app, it "creates" the tables...it should only do it once.  Avar & Bvar are returning "1" every time
2. The next tx.executeSql("INSERT...is not triggering (that I can tell).
3. Getting general error from isRegistered(...errorCB...); - which I believe is coming from function isTableExists()
Error processing SQL: 0, a statement error callback did not return false: no such table: myInfo (code 1): , while compiling: SELECT \* FROM myInfo
###### 

Error Code : 1 (SQLITE_ERROR)
Caused By : SQL(query) error or missing database.
    (no such table: myInfo (code 1): , while compiling: SELECT \* FROM myInfo)
###### 

*/
## Console messages:

```
OPEN database: myapp.db  
CT: registering tables  
new transaction is waiting for open operation  
DB opened: myapp.db  
Avar: 1 
CT: CHECKING myapp table  
Bvar: 1  
A1 Success: {"rows":{"length":0},"rowsAffected":0} 
B1 Success: , {"rows":{"length":0},"rowsAffected":0} 
Error processing SQL: 0, a statement error callback did not return false: no such table: myInfo (code 1): , 
while compiling: SELECT * FROM myInfo
################################################################# 
Error Code : 1 (SQLITE_ERROR) Caused By : SQL(query) error or missing database.     (no such table: myInfo (code 1): , while compiling: SELECT
* FROM myInfo)
#################################################################

```
## Code:

```
function isRegistered() {
  var db = window.sqlitePlugin.openDatabase({name: "myapp.db"});
  console.log("CT: registering tables") ;
  db.transaction(checkTable, errorCB,function(){
    console.log("Success") ;
  });
}

function isTableExists(tx, tableName, callback) {
  tx.executeSql('SELECT * FROM '+tableName, [], function(tx, resultSet) {
    if (resultSet.rows.length <= 0) {
        callback(false,0);
    } else {
        callback(true,2);
    }
  }, function(err) {
      callback(false,1);
  })
};

function checkTable(tx) {
  // create table
  isTableExists(tx, "myInfo", function(status,avar) {
    console.log("Avar: "+avar) ;
    if (!status) {
      console.log("CT: CHECKING myInfo") ;
      tx.executeSql("CREATE TABLE IF NOT EXISTS myInfo ('id' INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE,'name' VARCHAR,'phone' VARCHAR,'email' VARCHAR,'fbID' VARCHAR,'fbToken' VARCHAR,'regCode' VARCHAR)",[],function(tx,res) {
        console.log("A1 Success: "+JSON.stringify(res)) ;
        tx.executeSql('INSERT INTO myInfo VALUES (?,?,?,?,?,?,?)',[0,"","","","","",""],function(res) {
          console.log("A2 Success: " + JSON.stringify(res) + " -- probably 1");
        }, function(err) {
          console.log("A2 fail: "+JSON.stringify(err)) ;
        }) ;
      },function(err) {
        console.log("A1 Fail : "+ JSON.stringify(err)) ;
      }) ;
    }
  }, function(err){
    console.log("A Fail: "+JSON.stringify(err)) ;
  }) ;
  isTableExists(tx, "mySettings", function(status,bvar) {
      console.log("Bvar: "+bvar) ;
      if (!status) {
      tx.executeSql('CREATE TABLE IF NOT EXISTS "mySettings" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE,"vendor" VARCHAR,"display" VARCHAR,"aggregate" VARCHAR)',[],function(tx,res) {
        console.log("B1 Success: , "+ JSON.stringify(res)) ;
        var vendors = ["General","Twist","Unter","Sobi"] ;
        for (var n=0;n<vendors.length;n++) {
          tx.executeSql('INSERT INTO mySettings (id,vendor,display,aggregate) VALUES (?,?,?,?);',["","","",""],function(res) {
            console.log("B2 Success: " + n + " : " + JSON.stringify(res)) ;
          },function(err) {
            console.log("B2 Fail: "+JSON.stringify(err)) ;
          }) ;
        }        
      },function(err) {
          console.log("B1 Fail: "+JSON.stringify(err)) ;
      }) ;
    }
  }, function(err){
    console.log("B Fail: "+JSON.stringify(err)) ;
  }) ;
}
```

**Comment by brody2consult** [comment]
Unfortunately I cannot make much sense of your question. First, which platform(s) do you see the issue with (Android/iOS)?

I find your code really hard to follow. I wish you would have used Markdown Javascript highlighting and proper indentation.

**Comment by rolinger** [comment]
currently, I am only testing on Android.  I will try and clean up the code....ok, just did, yeah, that looks a lot better now.  sorry about that.

**Comment by brody2consult** [comment]
Thanks! I would suggest another general approach:

First start with a really simple test program like you can find in README.md.

Then make a table such as SchemaVersion, with a single row and column with a version number that starts with something like 1. It would be an easier way to keep track of whether or not you have created the tables and make future upgrading of your database tables more smooth.

**Comment by rolinger** [comment]
I tried the sample and it is working.  but I am trying to nest INSERTs inside of CREATE TABLE IF NOT EXIST success-callbacks...so only INSERT if the table needs to be created.  

As I look at the sample code from README.md that is working I am realizing there are somethings and relationships I am not fully understanding.  db.transaction makes a call to two tx.executeSql, but then also to a db.executeSql - what is the difference between the tx.executeSql and db.executeSql here?  Then later you have a another db.transaction nested inside of tx.executeSql?  Whats the difference with these three and where to use which?

Obviously I am confused here.  I have been pouring over readmes/howtos/docs for days and am just running in circles now

**Comment by brody2consult** [comment]
I documented that there are (now) three ways to execute SQL:
1. single-statement "transactions" using `db.executeSql`
2. batch sql queries using `db.sqlBatch`
3. standard asynchronous transactions (using db.transaction with a callback that can execute tx.executeSql)

Using db.transaction with tx.executeSql (3) is according to the Web SQL DRAFT standard API and would also work for a database opened using window.openDatabase instead of this plugin.

The single-statement "transactions" using db.executeSql (1) was originally introduced to support SQLite PRAGMA statements which should not be executed within a failure-safe transaction. The batch sql queries (2) was introduced to provide a simpler API for populating the database with data from sources such as a sqlite dump or an AJAX server.

Considering that Web SQL was documented much more than this plugin, I recommend that you start by playing with simple Web SQL in a browser and try things one step at a time until it makes more sense to you. In principle, almost anything that works on a Web SQL database object _should_ work with this plugin otherwise it should be reported as a bug.

**Comment by brody2consult** [comment]
Also please do not hesitate to ask if you have any more specific questions about the API of either Web SQL or this plugin. The purpose is to make it faster, NOT slower, for people to get started and working with databases. I would appreciate any suggestions you may have about the documentation as well.

In general, I have made major efforts to get the existing documentation to help people understand the purpose and capabilities of this project, understand the API, and get started as quickly as possible. In addition I have made efforts to document the known bugs, major pitfalls, and other shortcomings that may cause unnecessary confusion and time costs for developers getting started. Please let me know if you see any more roadblocks or other pitfalls that are not so obvious for new app developers.
